### DISCUSSION TITLE
http://genmon.local:8000 not working

### DISCUSSION THREAD

**Discussion by MarkKananen** [jgyates/genmon#1149] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).

### Version

V1.19.02

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have submitted logs

### What is/are your question(s)?

Why isn't the web interface working?

### Additional information

For the life of me I cannot get this working.  I have and older generac evolution.  I ordered a new pi zero w, and the hat from: https://pintsize.me/store/ols/products/generator-hats-and-phat/v/PGI-PHAT
Not actually sure I have the correct HAT.

I followed the instructions or genmon setup. The hat has constant blue light and it would appear the py scripts cannot communicate with it.  In genmon.conf I tried both serial0 and serial1.  /var/log/genmon.log:

mark@genmon:/var/log $ cat genmon.log
2024-07-21 17:05:23,056 : New version detected: Old = V1.19.03, New = V1.19.03
2024-07-21 17:05:23,151 : Selected Controller: generac_evo_nexus
2024-07-21 17:05:23,286 : Error opening modbus device: Error opening modbus device. : generac_evolution.py:325
2024-07-21 17:05:23,288 : Possible cause is invalid serial port specified.
2024-07-21 17:12:08,025 : Selected Controller: generac_evo_nexus
2024-07-21 17:12:09,124 : GenMon Loaded for site: SiteName using python 3.9: VEnv: False
2024-07-21 17:12:41,370 : WARNING: Initialization not complete after 30 seconds, possible communication failure. Check cabling.
2024-07-21 17:17:26,934 : Selected Controller: generac_evo_nexus
2024-07-21 17:17:27,943 : GenMon Loaded for site: SiteName using python 3.9: VEnv: False
2024-07-21 17:18:00,192 : WARNING: Initialization not complete after 30 seconds, possible communication failure. Check cabling.
2024-07-23 06:41:57,561 : Generator Monitor Shutdown
2024-07-23 06:43:24,168 : Selected Controller: generac_evo_nexus
2024-07-23 06:43:25,182 : GenMon Loaded for site: SiteName using python 3.9: VEnv: False
2024-07-23 06:43:57,429 : WARNING: Initialization not complete after 30 seconds, possible communication failure. Check cabling.

Granted, I do not have it plugged into the generator yet, I expected to see a UI before taking it outside.  At least the videos linked from the wiki indicate it should have the web interface available.

Any guidance, before I return everything and give up?  I've tried two micro SD cards, reinstalled several times...same results every time.

Thanks
Mark

**Comment by skipfire** [comment]
There isn't a wrong HAT, the hardware is the same for all supported controllers.

Did you try the serial loopback test while setting up?

**Comment by jgyates** [comment]
Juding only by the log output you show above it looks like you either have the wrong serial port selected, your serial port is not properly setup or your serial device is not responding.

You mentioned that you have not connected this to a generator yet. In that case it may take a few minutes to load the UI as the web interface will not load until an attempt has been made to read the generator registers. If you are not connected to the generator then this will result in numerious timeout errrors so if you you are not connected to the generator it may take a couple of minutes to bring up the UI with this many timeout errors.

I have not receveid any logs, I am guessing because you have not configured outbound email and can not pull up the web interface. In your message above you indicated that you did send logs so I am assuming that was unintentional.

What is the output of this command on the pi?

       ls -la /dev/ser*

You wrote "I followed the instructions or genmon setup." Did you run the instructions here: https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software

The script ./genmonmaint.sh asks questions and setups the serial port based on your response.

**Reply by MarkKananen** [reply]
mark@genmon:/var/log $  ls -la /dev/ser*
lrwxrwxrwx 1 root root 7 Jul 21 17:08 /dev/serial0 -> ttyAMA0
lrwxrwxrwx 1 root root 5 Jul 21 17:08 /dev/serial1 -> ttyS0

ok.  So loopback fails when run without jumping the pins.  It succeeds when jumping them.  Here is the startup:

mark@genmon:~/genmon $ bash startgenmon.sh stop
Stopping genmon python scripts
Stopping....
Stopping genserv.py
kill: (27511): No such process
Stopping genmon.py
mark@genmon:~/genmon $ cd OtherApps/
mark@genmon:~/genmon/OtherApps $ python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5b0
Serial port opened
write data: sent test string
waiting to received data....
PASSED! Loopback successful
mark@genmon:~/genmon/OtherApps $ bash startgenmon.sh start
bash: startgenmon.sh: No such file or directory
mark@genmon:~/genmon/OtherApps $ cd ..
mark@genmon:~/genmon $ cd ..
mark@genmon:~ $ cd genmon
mark@genmon:~/genmon $ bash startgenmon.sh start
Starting genmon python scripts
Warning: required library flask not installed. Attempting to install....
Starting....
Starting /home/mark/genmon/genmon.py
Starting /home/mark/genmon/genserv.py
mark@genmon:~/genmon $

**Reply by skipfire** [reply]
Then serial0 is the correct port to use.

**Comment by MarkKananen** [comment]
is there a way to try the loopback test without buying anything?

**Reply by jgyates** [reply]
being able to use the URL genmon.local may be a function of your router, have you tried using the IP address of the pi instead of genmon.local?

Here are alagotor clips:

https://www.amazon.com/Alligator-Electrical-21-5inch-Connection-Experiment/dp/B0995KJWR5/ref=asc_df_B0995KJWR5/?th=1

If you have something like this you can use it for a loopback test, however you may just want to try plugging it in to your controller first. So far you do not have any data that suggest your hardware is not working since you have not connected anything to the controller yet. The errors you are getting in your log are consistent with either selecting the wrong serial port and / or not having the pi connected to the controller.

If you show the output of this command:

      ls -la /dev/ser*

we can validate the serial port is active and determine which port you should use (you mentioned changing it to serial1, which is probably not correct)

The loopback test will only validate the serial port, it has nothing to do with the web interface. I would suggest the following:

1) If you have run the setup described here:

https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software

2) make sure you serial port in /etc/genmon/genmon.conf is set to /dev/serial0 . Do not edit the file in the /home/pi/genmon folder as this will do nothing to genmon

3) connect the serial cable and pi to the controller and assuming you selected genmon to start on boot you should be able to see results (assuming you are trying to access the pi at the right web address.

**Reply by skipfire** [reply]
I have found the scissors on a Swiss army knife work well for the loopback test as well. It is the two right pins on the top row directly on the HAT Molex plug. Do not do if barrel or Molex are connected.

**Reply by MarkKananen** [reply]
so the serial seems to be working as the serialtest.py fails without jumpers and succeeds with it.  I still cannot get anything in the browser.

**Reply by MarkKananen** [reply]
And using the ipaddress does not work either: http://192.168.0.225:8000

**Comment by MarkKananen** [comment]
There might be an error in the install.  It is installed to /home/mark/genmon as there was no /home/pi directory.

**Reply by MarkKananen** [reply]
I linked /home/mark to /home/pi, so that should not be an issue now.  This is all that is in the log:
2024-07-23 15:44:28,950 : WARNING: Initialization not complete after 30 seconds, possible communication failure. Check cabling.
2024-07-23 15:56:10,133 : Generator Monitor Shutdown
2024-07-23 15:57:41,239 : Selected Controller: generac_evo_nexus
2024-07-23 15:57:42,200 : GenMon Loaded for site: SiteName using python 3.9: VEnv: False
2024-07-23 15:58:14,454 : WARNING: Initialization not complete after 30 seconds, possible communication failure. Check cabling.

**Comment by MarkKananen** [comment]
It would appear there is no web server running:
mark@genmon:/var/log $ curl 192.168.0.225:8000
curl: (7) Failed to connect to 192.168.0.225 port 8000: Connection refused
mark@genmon:/var/log $ curl localhost:8000
curl: (7) Failed to connect to localhost port 8000: Connection refused
mark@genmon:/var/log $ curl 127.0.0.1:8000
curl: (7) Failed to connect to 127.0.0.1 port 8000: Connection refused
mark@genmon:/var/log $ curl kananen.com
mark@genmon:/var/log $ curl www.kananen.com
<!DOCTYPE html>
<html>
...

**Reply by MarkKananen** [reply]
Which is maybe consistent with this, seen during the start:
Warning: required library flask not installed. Attempting to install....

**Reply by MarkKananen** [reply]
That is about as far as I can debug this as I don't know much about python.  But for some reason FLASK is not getting installed and I do not see a way to get a reason why out of it.

**Reply by jgyates** [reply]
did you run the install script genmonmaint.sh -i? This is the scrip that should be run before starting genmon. 

If you did not then the loader program genloader.py will attempt to load the missing libaries but it will take a LONG time. It is not recommended to run genmon without running gnemonmaint.sh first.

When you do run genmonmaint.sh -i the install script will show any errors so if you are having an issue installing flask then this script will show the install errors.

If you did run genmonmaint.sh -i before then show the output of both /var/log/genserv.log and /var/log/genloader.log

**Reply by MarkKananen** [reply]
mark@genmon:~/genmon $ bash genmonmaint.sh -i
The operating system is: Raspbian GNU/Linux
This script will install libraries needed by genmon. This script assumes you have already downloaded the genmon project via 'git'. This script will optionally copy the config files to the configuration directory. This script will not test the serial port. This script requires internet access to download the needed libraries. Press any key to continue.
Installing genmon package requirements....
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: Flask in /usr/lib/python3/dist-packages (from -r /home/mark/genmon/requirements.txt (line 1)) (1.1.2)
Requirement already satisfied: configparser in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 2)) (7.0.0)
Requirement already satisfied: pyserial in /usr/lib/python3/dist-packages (from -r /home/mark/genmon/requirements.txt (line 3)) (3.5b0)
Requirement already satisfied: crcmod in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 4)) (1.7)
Requirement already satisfied: pyowm==2.10.0 in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 5)) (2.10.0)
Requirement already satisfied: pytz in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 6)) (2024.1)
Requirement already satisfied: pyopenssl in /usr/lib/python3/dist-packages (from -r /home/mark/genmon/requirements.txt (line 7)) (20.0.1)
Requirement already satisfied: twilio in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 8)) (9.2.3)
Requirement already satisfied: chump in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 9)) (1.6.0)
Requirement already satisfied: paho-mqtt==1.6.1 in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 10)) (1.6.1)
Requirement already satisfied: smbus in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 11)) (1.1.post2)
Requirement already satisfied: pysnmp in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 12)) (6.2.4)
Requirement already satisfied: ldap3 in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 13)) (2.9.1)
Requirement already satisfied: pyasn1==0.4.8 in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 14)) (0.4.8)
Requirement already satisfied: pyotp in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 15)) (2.9.0)
Requirement already satisfied: psutil in /usr/lib/python3/dist-packages (from -r /home/mark/genmon/requirements.txt (line 17)) (5.8.0)
Requirement already satisfied: spidev in /usr/lib/python3/dist-packages (from -r /home/mark/genmon/requirements.txt (line 18)) (3.5)
Requirement already satisfied: bleson in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 19)) (0.1.8)
Requirement already satisfied: fluids in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 20)) (1.0.25)
Requirement already satisfied: voipms==0.2.5 in /usr/local/lib/python3.9/dist-packages (from -r /home/mark/genmon/requirements.txt (line 21)) (0.2.5)
Requirement already satisfied: RPi.GPIO in /usr/lib/python3/dist-packages (from -r /home/mark/genmon/requirements.txt (line 22)) (0.7.0)
Requirement already satisfied: requests<3,>=2.20.0 in /usr/local/lib/python3.9/dist-packages (from pyowm==2.10.0->-r /home/mark/genmon/requirements.txt (line 5)) (2.32.3)
Requirement already satisfied: geojson<3,>=2.3.0 in /usr/local/lib/python3.9/dist-packages (from pyowm==2.10.0->-r /home/mark/genmon/requirements.txt (line 5)) (2.5.0)
Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.9/dist-packages (from requests<3,>=2.20.0->pyowm==2.10.0->-r /home/mark/genmon/requirements.txt (line 5)) (3.3.2)
Requirement already satisfied: certifi>=2017.4.17 in /usr/lib/python3/dist-packages (from requests<3,>=2.20.0->pyowm==2.10.0->-r /home/mark/genmon/requirements.txt (line 5)) (2020.6.20)
Requirement already satisfied: idna<4,>=2.5 in /usr/lib/python3/dist-packages (from requests<3,>=2.20.0->pyowm==2.10.0->-r /home/mark/genmon/requirements.txt (line 5)) (2.10)
Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/lib/python3/dist-packages (from requests<3,>=2.20.0->pyowm==2.10.0->-r /home/mark/genmon/requirements.txt (line 5)) (1.26.5)
Requirement already satisfied: scipy>=0.9.0 in /usr/local/lib/python3.9/dist-packages (from fluids->-r /home/mark/genmon/requirements.txt (line 20)) (1.13.1)
Requirement already satisfied: numpy>=1.5.0 in /usr/local/lib/python3.9/dist-packages (from fluids->-r /home/mark/genmon/requirements.txt (line 20)) (2.0.1)
Requirement already satisfied: pysmi<2.0.0,>=1.3.0 in /usr/local/lib/python3.9/dist-packages (from pysnmp->-r /home/mark/genmon/requirements.txt (line 12)) (1.4.4)
Requirement already satisfied: pysnmpcrypto<0.0.5,>=0.0.4 in /usr/local/lib/python3.9/dist-packages (from pysnmp->-r /home/mark/genmon/requirements.txt (line 12)) (0.0.4)
Requirement already satisfied: ply<4.0,>=3.11 in /usr/local/lib/python3.9/dist-packages (from pysmi<2.0.0,>=1.3.0->pysnmp->-r /home/mark/genmon/requirements.txt (line 12)) (3.11)
Requirement already satisfied: Jinja2<4.0.0,>=3.1.3 in /usr/local/lib/python3.9/dist-packages (from pysmi<2.0.0,>=1.3.0->pysnmp->-r /home/mark/genmon/requirements.txt (line 12)) (3.1.4)
Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.9/dist-packages (from Jinja2<4.0.0,>=3.1.3->pysmi<2.0.0,>=1.3.0->pysnmp->-r /home/mark/genmon/requirements.txt (line 12)) (2.1.5)
Requirement already satisfied: cryptography in /usr/lib/python3/dist-packages (from pysnmpcrypto<0.0.5,>=0.0.4->pysnmp->-r /home/mark/genmon/requirements.txt (line 12)) (3.3.2)
Requirement already satisfied: aiohttp>=3.8.4 in /usr/local/lib/python3.9/dist-packages (from twilio->-r /home/mark/genmon/requirements.txt (line 8)) (3.9.5)
Requirement already satisfied: PyJWT<3.0.0,>=2.0.0 in /usr/local/lib/python3.9/dist-packages (from twilio->-r /home/mark/genmon/requirements.txt (line 8)) (2.8.0)
Requirement already satisfied: aiohttp-retry>=2.8.3 in /usr/local/lib/python3.9/dist-packages (from twilio->-r /home/mark/genmon/requirements.txt (line 8)) (2.8.3)
Requirement already satisfied: async-timeout<5.0,>=4.0 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/mark/genmon/requirements.txt (line 8)) (4.0.3)
Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/mark/genmon/requirements.txt (line 8)) (1.4.1)
Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/mark/genmon/requirements.txt (line 8)) (23.2.0)
Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/mark/genmon/requirements.txt (line 8)) (1.9.4)
Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/mark/genmon/requirements.txt (line 8)) (1.3.1)
Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/mark/genmon/requirements.txt (line 8)) (6.0.5)
Installing rpirtscts...
~/genmon ~/genmon
fatal: destination path 'rpirtscts' already exists and is not an empty directory.
make: 'rpirtscts' is up to date.
Finished installing rpirtscts.
~/genmon
Copy configuration files to /etc/genmon/? (y/n)?n
Not copying *.conf to /etc/genmon/
Setup the raspberry pi onboard serial port? (y/n)?n
Not setting up serial port
Done.
Start genmon on boot? (y/n)?n
Not updating crontab.
mark@genmon:~/genmon $

**Reply by MarkKananen** [reply]
I ran it again, looks like everything was there.  I do not have a  /var/log/genserv.log

**Reply by MarkKananen** [reply]
mark@genmon:~/genmon $ cat /etc/issue
Raspbian GNU/Linux 11 \n \l

mark@genmon:~/genmon $ python --version
Python 3.9.2

**Reply by MarkKananen** [reply]
Trying to start the server directly yields this:

mark@genmon:~/genmon $ python3 genserv.py
This program requires the Flask library. Please see the project documentation at https://github.com/jgyates/genmon.
Error: cannot import name 'escape' from 'jinja2' (/usr/local/lib/python3.9/dist-packages/jinja2/__init__.py)
mark@genmon:~/genmon $

**Comment by MarkKananen** [comment, marked as answer]
ok got it.  in requirements.txt change
Flask -> Flask>=2.2.2
and then everything started as normal

**Reply by jgyates** [reply]
I am glad it is working. Out of curiosity are you running a 32 or 64 bit version of the OS?

I have a 64 bit that runs flask 3.0.0 and a test 32 bit system that is running Flask 2.0.3 with no issues. I know sometimes I have seen packages break for relatively short periods of time (before the maintiners can fix them) but usually switching from 32 to 64 or vice versa will correct the issue.

The fix you provided will work just, just trying to get data incase this turns out to be an issue for others.

**Comment by MarkKananen** [comment]
I did the 32 but image.  Just kind of followed along the best I could with the YouTube video.

I’m excited to get it in the generator now.

Live, love, pick and strum,
-MrK

On Jul 23, 2024, at 7:03 PM, jgyates ***@***.***> wrote:

﻿

I am glad it is working. Out of curiosity are you running a 32 or 64 bit version of the OS?

I have a 64 bit that runs flask 3.0.0 and a test 32 bit system that is running Flask 2.0.3 with no issues. I know sometimes I have seen packages break for relatively short periods of time (before the maintiners can fix them) but usually switching from 32 to 64 or vice versa will correct the issue.

The fix you provided will work just, just trying to get data incase this turns out to be an issue for others.
